src/activity_actor.cpp

```diff
@@ -8435,93 +8432,99 @@ void pulp_activity_actor::do_turn( player_activity &act, Character &you )
+            // Slow down the speed if corpse is acid
+            if( x_in_y( pulp_power, corpse.volume() / ( acid_corpse ? 20_ml : 50_ml ) ) ) {
+                corpse.inc_damage();
+                if( corpse.damage() == corpse.max_damage() ) {
+                    num_corpses++;
+                    corpse.set_flag( flag_PULPED );
                 }
+            }
 
-                // mixture of isaac clarke stomps and swinging your weapon
-                you.burn_energy_all( -pulp_effort );
-                you.recoil = MAX_RECOIL;
-
-                if( one_in( 4 ) ) {
-                    // Smashing may not be butchery, but it involves some zombie anatomy
-                    you.practice( skill_survival, 2, 2 );
+            if( x_in_y( pulp_power, corpse.volume() / 50_ml ) ) {
+                // Splatter some blood around
+                // Splatter a bit more randomly, so that it looks cooler
+                const int radius = mess_radius + x_in_y( pulp_power, 500 ) + x_in_y( pulp_power, 1000 );
+                const tripoint_bub_ms dest( pos + point( rng( -radius, radius ), rng( -radius, radius ) ) );
+                const field_type_id type_blood = ( mess_radius > 1 && x_in_y( pulp_power, 10000 ) ) ?
+                                                 corpse.get_mtype()->gibType() :
+                                                 corpse.get_mtype()->bloodType();
+                if( acid_corpse ) {
+                    here.add_splatter_trail( type_blood, pos, pos );
+                } else {
+                    here.add_splatter_trail( type_blood, pos, dest );
                 }
+            }
+            // mixture of isaac clarke stomps and swinging your weapon
+            you.burn_energy_all( -pulp_effort );
 
-                float stamina_ratio = static_cast<float>( you.get_stamina() ) / you.get_stamina_max();
-                moves += to_moves<int>( 6_seconds ) / std::max( 0.25f,
-                         stamina_ratio ) * you.exertion_adjusted_move_multiplier( act.exertion_level() );
-                if( stamina_ratio < 0.33 || you.is_npc() ) {
-                    you.set_moves( std::min( 0, you.get_moves() - moves ) );
-                    return;
-                }
-                if( moves >= you.get_moves() ) {
-                    // Enough for this turn;
-                    you.set_moves( you.get_moves() - moves );
-                    return;
-                }
+            if( one_in( 4 ) ) {
+                // Smashing may not be butchery, but it involves some zombie anatomy
+                you.practice( skill_survival, 2, 2 );
             }
-            corpse.set_flag( flag_PULPED );
         }
-        //Upon reach here, we have cleared one maptile
-        pos_iter = placement.erase( pos_iter );
     }
 
-    // If we reach this, all corpses have been pulped, finish the activity
+    if( current_pos_iter == placement.end() ) {
+        // no more locations to pulp
+        finish( act, you );
```

Comment:
You are not meant to call `finish()` explicitly. See the docs: https://github.com/CleverRaven/Cataclysm-DDA/blob/master/doc/PLAYER_ACTIVITY.md#termination.
I think simply setting `moves_left = 0` is appropriate.

Reply:
i copypasted the closest example for me, which is `wait_stamina_activity_actor`, which does call `finish()` directly; i'll check if assinging moves left does the same and use it instead